File: ABC/abc301-abc350/abc339/d/main.py
# -*- coding: utf-8 -*-


def main():
    import sys
    from collections import deque

    input = sys.stdin.readline

    n = int(input())
    s = [list(input().rstrip()) for _ in range(n)]
    pos = []

    for i in range(n):
        for j in range(n):
            if s[i][j] == "P":
                pos.append((i, j))

    # グリッドの状態を頂点とするグラフの最短経路問題に帰着 = BFS

    # 定数倍が重いので、座標のハッシュ値をキューに入れる
    # n進数に変換
    def to_hash(pos):
        p00, p01 = pos[0]
        p10, p11 = pos[1]

        value = p00
        value = value * n + p01
        value = value * n + p10
        value = value * n + p11

        return value

    def to_pos(hash):
        # to_hashとは逆順に
        p11 = hash % n
        hash //= n
        p10 = hash % n
        hash //= n

        p01 = hash % n
        hash //= n
        p00 = hash % n
        hash //= n

        pos = [(p00, p01), (p10, p11)]

        return pos

    n4 = n**4
    inf = 10**9

    dist = [inf] * n4
    id = to_hash(pos)
    dist[id] = 0

    q = deque([id])
    dxy = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (1, -1), (-1, 1), (1, 1)]
    dxy = dxy[:4]

    while q:
        cur_id = q.popleft()
        pos = to_pos(cur_id)
        d = dist[cur_id]

        # グラフを陽に持たず、1手先の状態を用意
        for dx, dy in dxy:
            npos = []

            for y, x in pos:
                nx = x + dx
                ny = y + dy

                # 条件を満たしていなければ元に戻す
                if nx < 0 or nx >= n or ny < 0 or ny >= n or s[ny][nx] == "#":
                    nx, ny = x, y

                npos.append((ny, nx))

            nid = to_hash(npos)

            if dist[nid] != inf:
                continue

            dist[nid] = d + 1
            q.append(nid)

    ans = inf

    for id, di in enumerate(dist):
        pos = to_pos(id)

        if pos[0] == pos[1]:
            ans = min(ans, di)

    if ans == inf:
        ans = -1

    print(ans)


if __name__ == "__main__":
    main()

File: ABC/abc301-abc350/abc339/d/main.rs
use std::collections::VecDeque;
use std::io::{self, BufRead};

fn to_hash(pos: &[(usize, usize); 2], n: usize) -> usize {
    let (p00, p01) = pos[0];
    let (p10, p11) = pos[1];

    let mut value = p00;
    value = value * n + p01;
    value = value * n + p10;
    value = value * n + p11;

    value
}

fn to_pos(hash: usize, n: usize) -> [(usize, usize); 2] {
    let mut hash = hash;
    let p11 = hash % n;
    hash /= n;
    let p10 = hash % n;
    hash /= n;
    let p01 = hash % n;
    hash /= n;
    let p00 = hash % n;

    [(p00, p01), (p10, p11)]
}

fn main() {
    let stdin = io::stdin();
    let mut input = stdin.lock().lines();

    let n: usize = input.next().unwrap().unwrap().trim().parse().unwrap();
    let mut s: Vec<Vec<char>> = Vec::new();
    let mut pos: Vec<(usize, usize)> = Vec::new();

    for i in 0..n {
        let line: Vec<char> = input.next().unwrap().unwrap().chars().collect();
        s.push(line);
        
        for j in 0..n {
            if s[i][j] == 'P' {
                pos.push((i, j));
            }
        }
    }

    let n4 = n * n * n * n;
    let inf = 1001001001;

    let mut dist = vec![inf; n4];
    let id = to_hash(&[pos[0], pos[1]], n);
    dist[id] = 0;

    let mut q = VecDeque::new();
    q.push_back(id);
    let dxy = [(-1, 0), (1, 0), (0, -1), (0, 1)];

    while let Some(cur_id) = q.pop_front() {
        let pos = to_pos(cur_id, n);
        let d = dist[cur_id];

        for &(dx, dy) in &dxy {
            let mut npos = [(0, 0); 2];

            for (i, &(y, x)) in pos.iter().enumerate() {
                let nx = (x as isize + dx) as usize;
                let ny = (y as isize + dy) as usize;

                if nx >= n || ny >= n || s[ny][nx] == '#' {
                    npos[i] = (y, x);
                } else {
                    npos[i] = (ny, nx);
                }
            }

            let nid = to_hash(&npos, n);

            if dist[nid] != inf {
                continue;
            }

            dist[nid] = d + 1;
            q.push_back(nid);
        }
    }

    let mut ans = inf;

    for (id, &di) in dist.iter().enumerate() {
        let pos = to_pos(id, n);

        if pos[0] == pos[1] {
            ans = ans.min(di);
        }
    }

    if ans == inf {
        ans = -1;
    }

    println!("{}", ans);
}